library/src/armory/export/criteria.py

```diff
@@ -0,0 +1,512 @@
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name: str, batch_idx, batch):
+        return chain_name in names
+
+    return _criterion
+
+
+def samples(indices: Sequence[int]) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches specific samples in the dataset, by
+    their global index (regardless of batch).
+
+    Example::
+
+        from armory.export import Exporter
+        from armory.export.criteria import samples
+
+        # For a batch size of 4:
+        #  - batch 0, sample 2
+        #  - batch 1, samples 0 and 1
+        #  - batch 2, samples 2 and 3
+        exporter = Exporter(criterion=samples([2, 4, 5, 10, 11]))
+
+    Args:
+        indices: Indices of samples within the dataset to be exported
+
+    Returns:
+        Export criterion function
+    """
+
+    def _criterion(chain_name, batch_idx: int, batch):
+        if len(indices) == 0:
+            return False
+        batch_size = len(batch)
+        return [
+            sample_idx
+            for sample_idx in range(batch_size)
+            if ((batch_idx * batch_size) + sample_idx) in indices
+        ]
+
+    return _criterion
+
+
+def _create_metric_criterion(comp, metric, threshold) -> Exporter.Criterion:
+    def _criterion(chain_name, batch_idx, batch):
+        val = metric(batch)
+        res = comp(val, threshold)
+        if type(res) == torch.Tensor:
+            if res.dim() == 0:
+                return res.item()
+            return res.nonzero().flatten().tolist()
+        return res
+
+    return _criterion
+
+
+def when_metric_eq(
+    metric: Callable[[Batch], Union[bool, float, torch.Tensor]],
+    threshold: Union[bool, float, torch.Tensor],
+) -> Exporter.Criterion:
+    """
+    Creates an export criterion that matches when a computed metric for a batch
+    or the samples within the batch is equal to a particular value.
+
+    Example::
+
+        import torch
+        from armory.data import DefaultTorchAccessor
+        from armory.export import Exporter
+        from armory.export.criteria import when_metric_eq
+
+        # Exports samples that have max score of exactly 5
+        def max_pred(batch):
+            return torch.tensor([
+                torch.max(p) for p in DefaultTorchAccessor().get(batch.predictions)
+            ])
+        exporter = Exporter(criterion=when_metric_eq(max_pred, 5))
+
+    Args:
+        metric: Callable that computes a metric for a batch. The return value
+            may be a single boolean or number, or it can be a tensor array of
+            the computed metric values for each sample in the batch.
+        threshold: Value the computed metric (either batchwise or samplewise)
+            must be equal to in order for the batch or samples to be exported
+
+    Returns:
+        Export criterion function
+    """
+    return _create_metric_criterion(lambda lhs, rhs: lhs == rhs, metric, threshold)
```

Comment:
This works for `int` values but might be problematic for `float` values -- it might be useful to consider an interface like [`math.isclose`](https://docs.python.org/3/library/math.html#math.isclose).